### DISCUSSION TITLE
Hunt strategy change for Maven repos is not how build systems which use Maven repositories behave

### DISCUSSION THREAD

**Discussion by JakeWharton** [renovatebot/renovate#25461] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

Placeholder value, please select the correct response from the dropdown

### Describe the problem

[This ticket](https://github.com/renovatebot/renovate/issues/22744) and [this PR](https://github.com/renovatebot/renovate/pull/24588) changed the strategy of resolving Maven dependencies from 'merge' to 'hunt'. The citation on the issue is that is how Maven (the build tool) purportedly behaves. This is describing how full Maven coordinate triples are resolved, not which versions are available in a repo, and as a result you are now going to miss dependency updates from artifacts which changed their Maven repo host.

Take, for example, the official Android Gradle plugin artifact whose coordinates are `com.android.tools.build:gradle`. Early on, Google published this to Maven Central as seen here: https://repo1.maven.org/maven2/com/android/tools/build/gradle/. Eventually, however, Google began publishing at a higher rate and higher quantity so they launched their own Maven repo hosted at maven.google.com. You can see the newer versions of the Android Gradle plugin artifacts here: https://maven.google.com/web/index.html#com.android.tools.build:gradle. Notice how the artifacts on Maven Central stop at 2.x.x and those on Google Maven start a 3.0.0.

The majority of artifacts an Android developer will consume come from Maven Central and so it usually appears first in the repo list. This ensures two things:
1. We're not sending Google data on which artifacts are in use by requesting them from Google first. And,
2. This speeds up your build because the majority of artifacts hit on the first repo. If the Google repo was first they would 404 and we would have to try again elsewhere.

As a result of this change, many repos stopped seeing updates. For example, [here's a PR](https://github.com/cashapp/paraphrase/pull/233) that got a new version before the above change was deployed. And [here's a Dependency Dashboard](https://github.com/cashapp/molecule/issues/141) whose job ran after the deploy and thus seems stuck on 8.1.1 despite 8.1.2 being out.

![Screenshot 2023-10-26 at 7 39 34 PM](https://github.com/renovatebot/renovate/assets/66577/c91aa6d1-9a6e-4aef-9518-7e8ecf3f1aab)

The Renovate job for that repo shows that it only checks Maven Central and never the Google Maven repo:
```
      {
        "datasource": "maven",
        "deps": [
          {
            "currentValue": "8.1.1",
            "currentVersion": "8.1.1",
            "datasource": "maven",
            "depName": "com.android.tools.build:gradle",
            "depType": "dependencies",
            "fileReplacePosition": 254,
            "fixedVersion": "8.1.1",
            "homepage": "http://tools.android.com/",
            "managerData": {
              "fileReplacePosition": 254,
              "packageFile": "gradle/libs.versions.toml"
            },
            "packageName": "com.android.tools.build:gradle",
            "registryUrl": "https://repo.maven.apache.org/maven2",
            "registryUrls": [
              "https://repo.maven.apache.org/maven2",
              "https://dl.google.com/android/maven2/",
              "https://plugins.gradle.org/m2/"
            ],
            "sourceUrl": "https://android.googlesource.com/platform/tools/base",
            "versioning": "gradle",
            "warnings": [],
            "updates": []
          },
 ⋮
DEBUG: Looking up com.android.tools.build:gradle in repository https://repo.maven.apache.org/maven2/
 ⋮
DEBUG: Found 76 new releases for com.android.tools.build:gradle in repository https://repo.maven.apache.org/maven2/
```

If I manually bump the version to 8.1.2, the build will successfully resolve the artifact. This is because Gradle will attempt to fetch it first from Maven Central and fail, it will then move on to Google Maven, my second configured repo, and attempt there.

Ultimately, when I read the [Maven build tool link](https://maven.apache.org/guides/mini/guide-multiple-repositories.html#repository-order) referenced in the original ticket I actually think this change is in error. Because that documentation is detailing the behavior when resolving a Maven coordinate _triple_ (i.e., a groupId and artifact with a version). It is not detailing how it resolves the `maven-metadata.xml` files which provide a list of versions of a groupId+artifactId pair.

You can reproduce this with a repo that contains a single `build.gradle` file with the following contents:
```gradle
apply plugin: 'java'

repositories {
  mavenCentral()
  google()
}

dependencies {
  implementation 'com.android.tools.build:gradle:8.1.1'
}
```
Note that this project builds (by running `gradle build` proving the resolution algorithm checks both repos).

I made a minimal reproducer for the Maven build tool as well. Put the following into a `pom.xml` in a repo:
```xml
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.example</groupId>
  <artifactId>example</artifactId>
  <version>1.0-SNAPSHOT</version>
  <repositories>
    <repository>
      <id>maven-central</id>
      <url>https://repo1.maven.org</url>
    </repository>
    <repository>
      <id>google-maven</id>
      <url>https://maven.google.com</url>
    </repository>
  </repositories>
  <dependencies>
    <dependency>
      <groupId>com.android.tools.build</groupId>
      <artifactId>gradle</artifactId>
      <version>8.1.1</version>
    </dependency>
  </dependencies>
</project>
```
This project also builds (using `mvn compile`), proving that Maven (the build tool) will check both repositories despite the first containing a set of artifacts matching the groupId+artifactId but not having a matching version.

Both of these projects when run with Renovate will not see the 8.1.2 upgrade. However, manually bumping the dependency to 8.1.2 will successfully build in both.

With this, I hope that I've convinced you that the linked commit above should be reverted, or at least changed to a strategy which honors full coordinate triples in the first Maven repo they're found in, but still checks all Maven repos for those sets of triples.

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by rarkins** [comment]
We may need a more complex approach than simply changing mergeStrategy from hunt to merge. One challenge was that Renovate in many cases was searching non-Maven Central repositories _unnecessarily_, leading to increased search times, increased load on that server, etc. What would be the solution in that case? e.g.

- In such cases users should use packageRules to manually override registryUrls to remove the second repo when it's not needed, or
- We need to make hunt/merge user-configurable, like registryUrls?
- In such cases users should configure their maven or gradle better so that it references the alternative registry only when necessary, and not for all?

Cc @Churro

**Reply by JavierSegoviaCordoba** [reply]
> One challenge was that Renovate in many cases was searching non-Maven Central repositories unnecessarily

How do you know it is unnecessary?

**Reply by rarkins** [reply]
The registries contained only company-proprietary packages but Renovate searched it for every single open source package too

**Reply by viceice** [reply]
same happens on nuget, but I'll submit a feature next week to support filtering registries by source mapping. maybe Gradle/ maven should adopt a similar feature.

maybe we can also add such a feature as general function inside renovate? 🤔

**Reply by JakeWharton** [reply]
Gradle supports filtering the lookups with include/exclude by groupId or regex: https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/repositories/RepositoryContentDescriptor.html

For example:
```groovy
repositories {
  maven {
    url 'https://.....'
    content {
      includeGroup('com.example')
    }
  }
  mavenCentral()
}
```
Despite being listed first, only artifacts whose groupId matches `com.example` will be fetched from the repo. All other artifact coordinates will skip resolution in the first repo and go to the second.

Is that what you're talking about?

**Reply by viceice** [reply]
@Churro we should extract those and filter registries, this will reduce requests when properly configured.

**Reply by rarkins** [reply]
@JakeWharton if Gradle already supports that syntax, and then we parse/interpret it correctly in Renovate, I think that's a very valid answer for any one who has a complaint that Renovate is looking up their private registry for open source components. Does anyone know if Maven has something similar?

Also remember: Renovate can allow users to control registryUrls using packageRules, so this would be resolvable through _config_ regardless.

**Reply by Churro** [reply]
Agreed, adding support for the content descriptor syntax makes sense in any case. Still, it seems to me that the originally described situation of packages being moved between registries, would go unnoticed unless is aware and uses `includeGroup` (or a related construct) to prioritize. `packageRules` already allow for this in renovate.

Some time ago, I started working on https://github.com/renovatebot/renovate/issues/14208 and was aiming for a "universal" solution 
that would cover all content descriptors in `maven { content { ... } }` and `exclusiveContent { filter { ... } }`. I'll take this as an incentive to take another look - it may not cover all possibilities but at least the most common ones.

**Reply by JavierSegoviaCordoba** [reply]
Is there any workaround until the final implementation is done?

**Reply by gabrielittner** [reply]
As an alternative, could the `hunt` strategy be changed to instead of just checking that the package exists to also check that the version exists? If it would do that renovate would check maven central for Android Gradle Plugin and then continue searching for it. It would still have this issue for packages that switched registries/repositories when someone is using a version from before the switch (for Android Gradle plugin it would be versions < 3.x) but that is a one time issue and it will work again after updating manually once.

**Reply by rarkins** [reply]
As a workaround you can switch the order of registries using packageRules.registryUrls

**Reply by JavierSegoviaCordoba** [reply]
Is it not possible to disable the `hunt` strategy? A toggle config would be great.

I can't see how changing the order can solve the issue, as it can solve that one artifact is not being updated, but that fix can break other artifacts. And, it can mean checking/creating rules for a lot of artifacts...

**Reply by rarkins** [reply]
It shouldn't be too common where you have packages found in _both_ registries and the second one is the one you care about. In such cases then you can target those dependencies using packageRules and registryUrls.

Configuring registryStrategy is not supported

**Reply by JavierSegoviaCordoba** [reply]
Yeah, I agree with that, but even if it is not common, you can't know you are not missing some package if you don't check for any package in your dependencies manually, which can be painful.

**Reply by rarkins** [reply]
To be clear: it's a workaround, not a solution

**Reply by thomaszub** [reply]
I stumbled across the same issue and can give some use case in Maven where multiple repositories are used:
It is quite typical to separate Release-Versions and Snapshot-Versions into different groups in Nexus (an common artifact repository for Java), each then having a different URL, e.g. 

 https://nexus.mycorp.de/nexus/content/groups/releases-group
 https://nexus.mycorp.de/nexus/content/groups/snapshots-group

See as an example https://oss.sonatype.org/#view-repositories

Then two registries are configured in the Maven settings, one for Release-Version and for Snapshot-Versions. Now Maven checks both groups for versions and picks the appropriate artifact depending on the full coordinates. With the hunt strategy, potential versions are missing.

**Reply by rarkins** [reply]
We will revert this change back to how it was in v36 and provide instructions for those who wish to minimize requests to internal servers

**Reply by thomaszub** [reply]
@rarkins: When will the v36 approximately be released? Asking for a another developer who I am supporting with Renovate.

**Reply by rarkins** [reply]
It's v38 next and will be before end of year, maybe end of November

**Reply by rarkins** [reply, marked as answer]
We plan to release it in v37 instead: #26013

**Reply by eygraber** [reply]
Does v37 correlate with `renovateVersion` in the job logs? If so I'm still seeing the issue where `com.android.tools.build:gradle` matches in maven central and never finds the versions from maven.google.com ([job log](https://developer.mend.io/github/eygraber/compose-placeholder/-/job/018c22b1-7e44-7f68-9504-31c8fe4866a6).

**Reply by rarkins** [reply]
If you look in the linked PR you'll see:

![image](https://github.com/renovatebot/renovate/assets/6311784/c07fb2ba-c56c-4822-a163-ab5b6ac2ef78)

If you look in a recent Renovate app log you'll see:

![image](https://github.com/renovatebot/renovate/assets/6311784/61da61e8-40de-4120-9972-88a48b351bd5)

In other words, the app is not yet running the latest so needs to be updated before the functionality change is "live". This should hopefully happen on Sunday

**Comment by jonesbusy** [comment]
Hunt strategy was also causing conflict with release/snapshot repositories order : https://github.com/renovatebot/renovate/discussions/25294

Thanks for the revert

**Comment by aikebah** [comment]
@rarkins I recently discovered this ticket after observing item 1. of the opening post in one of my debug logs (self-hosted renovate). I don't see that addressed in the context of this ticket. Any plans on fixing up hammering Maven Central with lots of useless requests? The IS NO index.html, there is an HTML-page 'directory listing' served if you browse the folder location, so the `index.html` should not be added to the URL.

E.g. 
`https://repo.maven.apache.org/maven2/com/android/tools/build/gradle/` lists the versions of the `com.android.tools.build:gradle` hosted on maven central and is the HTML that should be parsed when you don't want to trust on maven metadata, but the current code will instead try to retrieve and parse `https://repo.maven.apache.org/maven2/com/android/tools/build/gradle/index.html` which will always yield a HTTP-404

That is... to resolve item 1. at
https://github.com/renovatebot/renovate/blob/a13090c12e7f3602ea7dcc8ba32960fb75425e60/lib/modules/datasource/maven/index.ts#L146 
`'index.html'` should be changed to `''` to retrieve the package index page from Maven Central.

**Reply by rarkins** [reply]
PR welcome. Otherwise please start a new discussion with your idea

**Comment by rarkins** [comment]
Locking this as I'd like any new discussions on the topic to be in a new github discussion
